### ISSUE TITLE
Discard changes does not delete new files when Recycle settings altered

### ISSUE THREAD

**Issue by BasmanovDaniil** [desktop/desktop#4194]
## Description
The old GitHub for Windows was able to discard any changes made to a repository, you could mess with files all you like, then hit discard and have a pristine clean repo. Now, if you have created any new files, you have to use console commands or Explorer to get rid of them.

## Version
* GitHub Desktop: 1.1.0
* Operating system: Windows 10 Pro

## Steps to Reproduce
1. Open any repository in GitHub Desktop
2. Create a new file
3. Try to discard it

### Expected Behavior
The file is dead

### Actual Behavior
It is still there

**Comment by iAmWillShepherd** [comment]
Thanks for taking the time to open an issue! 🙇 

Here's what I get using the steps given:
![2018-03-07 09 52 53](https://user-images.githubusercontent.com/1715082/37102440-742731b2-21ed-11e8-9454-2c2d4af336da.gif)

Could you attach your logs please so I can make sure no errors are occurring in the background on your end?

**Comment by BasmanovDaniil** [comment]
Well, here is how it looks on my computer:
![](https://monosnap.com/file/cV1PbgGVCiJimOBU89Fkx4KoRMwJD4.png)

> 2018-03-07T16:31:00.566Z - info: [ui] [AppStore] loading 8 repositories from store
> 2018-03-07T16:31:00.568Z - info: [ui] [AppStore] found account: BasmanovDaniil (Daniil Basmanov)
> 2018-03-07T16:31:01.570Z - info: [ui] launching: 1.1.0 (Windows 10.0.16299)
> 2018-03-07T16:31:01.570Z - info: [ui] execPath: 'C:\Users\username\AppData\Local\GitHubDesktop\app-1.1.0\GitHubDesktop.exe'

**Comment by BasmanovDaniil** [comment]
The path to the repository is `C:\Code\Test`, nothing fancy.

**Comment by iAmWillShepherd** [comment]
@BasmanovDaniil, I was able to reproduce on Windows only.

**Comment by shiftkey** [comment]
Moving this into 1.1.x as this appears to be a recent regression.

**Comment by shiftkey** [comment]
@iAmWillShepherd I've not been able to reproduce this in a test harness. Assigning to you to figure out what the root cause of this is.

**Comment by iAmWillShepherd** [comment]
@BasmanovDaniil, I was not able to reproduce this issue again. Does this behavior happen consistently?

**Comment by BasmanovDaniil** [comment]
@iAmWillShepherd, Yes, I thought it was a feature until now. Actually, I use it as a feature 😄. I just became annoyed enough over the past month to open an issue.

**Comment by BasmanovDaniil** [comment]
I behaves the same way in of all my repositories, github and non-github, local and remote.

**Comment by NiklasBr** [comment]
I encountered this (or a very similar) issue on macOS 10.12.6 (16G1408) today.

No matter what I do (restart, app or computer) the changes to files will not be discarded. The log writes the line:

`2018-06-15T07:49:05.019Z - info: [ui] No submodules found. Skipping "git submodule status"`

Every time I try to discard any changes.

[2018-06-15.desktop.production.log](https://github.com/desktop/desktop/files/2105219/2018-06-15.desktop.production.log)

**Comment by shiftkey** [comment]
@NiklasBr that message is just part of our check for any submodules and can be ignored.

Can you attach the output of this command and indicate which files you are trying to discard?

```
$ git status --porcelain=2 --untracked=all
```

**Comment by NiklasBr** [comment]
@shiftkey I'm trying to discard all of these:

``` shell
$ git status --porcelain=2 --untracked=all
1 .M N... 100644 100644 100644 73984fcef394abc2441474e72fbaac017867b035 73984fcef394abc2441474e72fbaac017867b035 sites/all/libraries/PHPMailer/phpmailer/LICENSE
1 .M N... 100644 100644 100644 8488ca6dc161e292b016578729c414db5663c427 8488ca6dc161e292b016578729c414db5663c427 sites/all/libraries/PHPMailer/phpmailer/get_oauth_token.php
1 .M N... 100644 100644 100644 45138343361d75b7157864da44dd1267382b71a6 45138343361d75b7157864da44dd1267382b71a6 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-am.php
1 .M N... 100644 100644 100644 917c9c95b1eb2f01ed930e335c40109719f98145 917c9c95b1eb2f01ed930e335c40109719f98145 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-ar.php
1 .M N... 100644 100644 100644 903dfaa43f41341f0670e1ee388c3eb40afbeb7e 903dfaa43f41341f0670e1ee388c3eb40afbeb7e sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-be.php
1 .M N... 100644 100644 100644 551205128ce14a2215b1a1f141cd118295b02722 551205128ce14a2215b1a1f141cd118295b02722 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-bg.php
1 .M N... 100644 100644 100644 8415e618352c5b7479cbe6b6856ae479892d38fa 8415e618352c5b7479cbe6b6856ae479892d38fa sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-cs.php
1 .M N... 100644 100644 100644 19b9579ede4a76ddfd9c76663ca89cfeae667588 19b9579ede4a76ddfd9c76663ca89cfeae667588 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-de.php
1 .M N... 100644 100644 100644 1d9061b3364f2d651ec5dd3ad1882333beafbf01 1d9061b3364f2d651ec5dd3ad1882333beafbf01 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-el.php
1 .M N... 100644 100644 100644 070477cea4120a3d00cdcfa70f50ba1fc3e12a79 070477cea4120a3d00cdcfa70f50ba1fc3e12a79 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-es.php
1 .M N... 100644 100644 100644 c7d137719c94ba9a817a78d73929ec2450dd056c c7d137719c94ba9a817a78d73929ec2450dd056c sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-fa.php
1 .M N... 100644 100644 100644 d48245545f0e729d373343fb20e0b76a363cefc8 d48245545f0e729d373343fb20e0b76a363cefc8 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-fr.php
1 .M N... 100644 100644 100644 e2c22537f7dc64cd28de23ac4f82053bba89120d e2c22537f7dc64cd28de23ac4f82053bba89120d sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-gl.php
1 .M N... 100644 100644 100644 4317627be791ad33b50ee919f583eb18d9404d4f 4317627be791ad33b50ee919f583eb18d9404d4f sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-he.php
1 .M N... 100644 100644 100644 2c1330c36c634cc92e1a648e66bfb97ab8e24cfd 2c1330c36c634cc92e1a648e66bfb97ab8e24cfd sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-hr.php
1 .M N... 100644 100644 100644 ce3c70478efaf988c6def08c5ba9d2f44f4343a1 ce3c70478efaf988c6def08c5ba9d2f44f4343a1 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-ja.php
1 .M N... 100644 100644 100644 3419ab6e96cf1e37138f0ca8f4da9d9b3a6c0f93 3419ab6e96cf1e37138f0ca8f4da9d9b3a6c0f93 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-ka.php
1 .M N... 100644 100644 100644 3f8c309b834601ffb5405c35b42def6cb7bd5df6 3f8c309b834601ffb5405c35b42def6cb7bd5df6 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-ko.php
1 .M N... 100644 100644 100644 06419c4608ce77abf8d82a6a8235f8d65a0831bd 06419c4608ce77abf8d82a6a8235f8d65a0831bd sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-lv.php
1 .M N... 100644 100644 100644 488b5bf2f88de3bd7c99ffbcc29201d954a433da 488b5bf2f88de3bd7c99ffbcc29201d954a433da sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-nl.php
1 .M N... 100644 100644 100644 b554589da4a9daaece4d777b14f8f88397aa9784 b554589da4a9daaece4d777b14f8f88397aa9784 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-pl.php
1 .M N... 100644 100644 100644 be6901427db672fb3f7d85b3f6f2afdd0e7ec40a be6901427db672fb3f7d85b3f6f2afdd0e7ec40a sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-pt.php
1 .M N... 100644 100644 100644 244b4f30fd4219a53501e404c47e5b451ff7f5bc 244b4f30fd4219a53501e404c47e5b451ff7f5bc sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-ro.php
1 .M N... 100644 100644 100644 c036594893cf9bba304da742df72487fc5d93ddf c036594893cf9bba304da742df72487fc5d93ddf sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-sk.php
1 .M N... 100644 100644 100644 3fb559df8da71f167b95afa79e0b8f41e1dcb173 3fb559df8da71f167b95afa79e0b8f41e1dcb173 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-sl.php
1 .M N... 100644 100644 100644 db163f7240bb3b44076f8db92c30124a633f35e5 db163f7240bb3b44076f8db92c30124a633f35e5 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-uk.php
1 .M N... 100644 100644 100644 310062bd163a0dfe2c81bff794530c81d616f298 310062bd163a0dfe2c81bff794530c81d616f298 sites/all/libraries/PHPMailer/phpmailer/language/phpmailer.lang-vi.php
1 .M N... 100644 100644 100644 4cf31f998bdfca120c04c4e954084e7baaf72d0e 4cf31f998bdfca120c04c4e954084e7baaf72d0e sites/all/libraries/PHPMailer/phpmailer/src/Exception.php
1 .M N... 100644 100644 100644 085eb78138047a2b201d131b925566a96f12f631 085eb78138047a2b201d131b925566a96f12f631 sites/all/libraries/PHPMailer/phpmailer/src/OAuth.php
1 .M N... 100644 100644 100644 3e904460b4567289684965cde848c083716a1487 3e904460b4567289684965cde848c083716a1487 sites/all/libraries/PHPMailer/phpmailer/src/POP3.php
1 .M N... 100755 100755 100755 bd70f329579a75983625a4cbbbdf2856a3ae7580 bd70f329579a75983625a4cbbbdf2856a3ae7580 sites/all/libraries/security-checker.phar
1 .M N... 100755 100755 100755 a8c96255f6af7a28a474622afb11c6f6b4df3d05 a8c96255f6af7a28a474622afb11c6f6b4df3d05 sites/all/libraries/squizlabs/php_codesniffer/src/Standards/Generic/Tests/Files/LineEndingsUnitTest.css
1 .M N... 100755 100755 100755 a2d3ed209351c60e994f494312a8ba7caa172a07 a2d3ed209351c60e994f494312a8ba7caa172a07 sites/all/libraries/squizlabs/php_codesniffer/src/Standards/Generic/Tests/Files/LineEndingsUnitTest.inc
1 .M N... 100755 100755 100755 49c58d8e051784b73e0feaf7c51c1b269fd3ddbc 49c58d8e051784b73e0feaf7c51c1b269fd3ddbc sites/all/libraries/squizlabs/php_codesniffer/src/Standards/Generic/Tests/Files/LineEndingsUnitTest.js
1 .M N... 100755 100755 100755 5373f2b882e30837c03d7445d676549e20024efb 5373f2b882e30837c03d7445d676549e20024efb sites/all/libraries/squizlabs/php_codesniffer/src/Standards/Squiz/Tests/Commenting/FileCommentUnitTest.inc
1 .M N... 100755 100755 100755 b870963be28830c847d2a25cfdc2390241629743 b870963be28830c847d2a25cfdc2390241629743 sites/all/libraries/squizlabs/php_codesniffer/src/Standards/Squiz/Tests/PHP/NonExecutableCodeUnitTest.inc
```

**Comment by schurigh** [comment]
I get this error, too. It seems to be new with my newly installed laptop, havent had this for months on my workstation before. Rightclick -> Discard all -> nothing happens and just one "No submodules found. Skipping "git submodule status"" for every try, no errors.
I will analyse this a bit deeper on Monday.

**Comment by msftrncs** [comment]
@billygriffin, I am not sure this has anything to do with submodules, unless I am unclear what that really means.

I just run in to this issue trying to clean up my sandbox repository after using UNDO during a merge conflict, per #6429.  I cannot revert the two 'new' files that were left over from the attempted merge.  Discarding a 'changed' file was no problem.

Oh Bingo …  https://github.com/desktop/desktop/pull/4703#issuecomment-403611839   That's the problem.  I am working on a USB stick … no recycling bin there.  This really is #5157.

I rebuilt with a check of the result from moveItemToTrash, the result was FALSE.  I think that means it failed.  That should probably thow an error.  I am thinking that IAppShell.moveItemToTrash needs to implement a little logic there, warn user that 'move to trash' failed and ask if it is okay to throw away instead.  The operation needs to continue, not fail with no reason.

**Comment by alexagranov7** [comment]
I'm running into the same issue. Windows 10, version 1.5.1.

**Comment by marianmelinte93** [comment]
Confirming this is still is an issue on version 2.3.1. Found the issue in my case to be the

**Comment by marianmelinte93** [comment]
Confirming this is still an issue on version 2.3.1 (Windows 10).
Found out that the problem, in my case, to be caused by setting the Recycle Bin to delete the files immediately, instead of moving them.

![recycle-bin](https://user-images.githubusercontent.com/13729459/75275219-dc792480-580c-11ea-8ab7-88ae14be0ca2.png)

**Comment by tierninho** [comment]
@magicmarvman Many thanks for the confirmation. I can reproduce that behavior as well! Hopefully we can get this fixed or at least warn users to alter their settings.

**Comment by Markus-Gurkcity** [comment]
Hi,
I have the same problem and tried the solution of @marianmelinte93 
I disabled the recycle bin for all drives and restart windows. but the bug exists. i can not discard changes the same way @BasmanovDaniil showed in his video.
any suggestions? 
Windows 10
GitHub Desktop 2.4.2
SSD as hard drive

**Comment by HedvigS** [comment]
(I'm also having this / a very similar issue, Github desktop: 2.5.3 Mac OS Catalina.)

**Comment by Crembotz** [comment]
I can confirm this is still happening, Windows 10 GitHub Desktop ver. 2.5.3

**Comment by HedvigS** [comment]
(It's happening again to me so I'm checking in on this issue thread again. Any news?)

**Comment by HedvigS** [comment]
Here's what it looks like when I try to do it. I'm on Mac OS Catalina 10.15.6 with GitHub desktop Version 2.5.5.

![Screen-Recording-2020-09-17-at-17 43 42](https://user-images.githubusercontent.com/5327845/93495371-56d6d680-f90e-11ea-8602-a54d1cb715ef.gif)

At this point I don't care about pride anymore, I'd be happy for someone to tell me that I'm doing it all wrong and how to correct it. I just don't want this problem anymore.

**Comment by tierninho** [comment]
@HedvigS Please attach your logs for the day this happened -- see the Help menu in Desktop. Thanks!

**Comment by IRampagingRhin0** [comment]
Yeah I just got this issue, I fetched the origin and one file got changed somehow (I haven't the means to even open the file). In the log files I get [2020-10-07T22:36:10.911Z - info: [ui] No submodules found. Skipping "git submodule status"]. Was there a fix?

**Comment by HedvigS** [comment]
Anything fun in that log that you noticed?

**Comment by leopucci** [comment]
Github 2.6.0
Windows 10 (19041.264) 64 bits

The error was occurring because the files were open by another program. 
After I closed the program, I could delete the files via github desktop.
The hint was trying to delete them via Explorer. It shows a message telling that the files are open by another program.

![image](https://user-images.githubusercontent.com/38509299/100227606-25771a00-2f00-11eb-9b3a-fa71322e7dfa.png)

If you guys put a message telling this to the user, problem solved. 
Regards,
Pucci

**Comment by marianmelinte93** [comment]
@leopucci, that doesn't seem to be the case for me.

GitHub 2.6.0
Windows 10 2004 (19041.630)

Here's an example where I can delete the file manually:
![Discard issue](https://user-images.githubusercontent.com/13729459/100238892-66d7eb80-2f39-11eb-9ad6-6bb310bd4191.gif)
And this is where I change the Recycle Bin setting to allow it to store the deleted files, and it works.
![Discard no issue](https://user-images.githubusercontent.com/13729459/100238951-79eabb80-2f39-11eb-84cb-610b99e0bd64.gif)

The same issue seems to happen when I try to remove the repository and also move it to Recycle Bin.
![Remove repository popup](https://user-images.githubusercontent.com/13729459/100239658-5411e680-2f3a-11eb-9295-eea14fe01209.png)
It does say "move ..  to Recycle Bin", but the Recycle Bin allows us to specify whether we want to store them temporarily or delete them permanently.
![Remove repository error](https://user-images.githubusercontent.com/13729459/100239339-f8475d80-2f39-11eb-987f-b654de494fc6.png)

Seeing that this issue is also present on MacOS, I'm not sure if it could be the same issue there. I'm not familiar with that OS.

**Comment by HedvigS** [comment]
I've had to delete my clone on my local machine and reclone the repos from origins 5 times in the last 24h because of this. Is there any progress?

**Comment by leopucci** [comment]
> I've had to delete my clone on my local machine and reclone the repos from origins 5 times in the last 24h because of this. Is there any progress?

To me, on windows, the problem was that files were open on another programs and also some programs that was inside the folder was running and generated files... One thing or another... Or both..
After I quit the programs that generated the files and was keeping them open. I could remove them. 
Maybe also a shell open on that folder... Try to investigate and test similar things... Maybe you can work it out as I did.

**Comment by HedvigS** [comment]
@leopucci Yes, I read that. I'm not on a windows machine and the file is not open in another program. I can try and see if there is something going on elsewhere where it's reading or writing to that folder, but I highly doubt it. I'll check, but I must say I suspect it's something else. But thanks.

**Comment by HedvigS** [comment]
I've already posted my log here in this thread. I'm having his issue again with other files and other branches. I can upload more logs if @tierninho wants.

**Comment by leopucci** [comment]
> @leopucci Yes, I read that. I'm not on a windows machine and the file is not open in another program. I can try and see if there is something going on elsewhere where it's reading or writing to that folder, but I highly doubt it. I'll check, but I must say I suspect it's something else. But thanks.

Have you tryed lsof?
`lsof +D /path/folder`

This will output open files on that dir. (I think. I´m not in linux right now. Just googled)

When you cant delete, try running this to check if there is some open file. 

There is also an option to check working directory of a program. 
`$ lsof -p 12136 | grep cwd`  
`nautilus 12136 saml  cwd    DIR              253,2    32768  10354689 /home/saml`

So you could find some process that is holding the folder (I think)

Release the Sherlock Holmes that exists inside you. :)

**Comment by HedvigS** [comment]
Well I couldn't wait around so I've just recloned to remove the problem. But sure, I can try that next time (which I'm sure will be soon enough given the project I'm working on).

I'm still not convinced it's an issue about open files, but I will investigate. 

From my perspective as a user of the desktop client, I'd prefer it if the path I'm currently being led on by the dialogue boxes made more sense.

**Comment by leopucci** [comment]
Also comes into my mind nautilus or some "windows explorer" folder visual
tool. Emacs.. vi... Git commands on the shell(?)

On the 90's I used to restart all X window system using ctrl + alt +
backspace. This will close all gui interface and programs(I guess) so then
you could login again and open only the github program and try to delete as
all gui programs will be closed.
This could give you some clue if it is a gui visual tool that is holding
everything.

Happy hunting!

Em ter, 26 de jan de 2021 18:44, Hedvig Skirgård <notifications@github.com>
escreveu:

**Comment by HedvigS** [comment]
Sure, okay. 

Did anyone at GitHub desktop client dev look at the log file from before?

**Comment by uberspeck** [comment]
I just ran into this issue today for the first time after using Github desktop for years. I just did a rebuild of my MacBook and a fresh install of Github Desktop refuses to discard changes. For that matter, `git checkout -- path/to/file` and `git restore path/to/file` don't seem to work either 😬

**Comment by leopucci** [comment]
> I just ran into this issue today for the first time after using Github desktop for years. I just did a rebuild of my MacBook and a fresh install of Github Desktop refuses to discard changes. For that matter, `git checkout -- path/to/file` and `git restore path/to/file` don't seem to work either 😬

Aren't those files open in some editor or running some program from the folder @uberspeck ? In my case I solved when I released the lock of running files and folders

**Comment by ted423** [comment]
Consider if the path and filename too long?

**Comment by tidy-dev** [comment]
Closed by #13904

Since upgrading electron in this past year, the logic around discard files was updated both in electron and in Desktop. I think this issue should be resolved now. If an error occurs when deleting files, it will give the option to delete them permanently. At the very least, this definitely handles the scenario where users recycle settings are altered.
